staging/src/k8s.io/apiserver/pkg/util/version/version.go

```diff
@@ -0,0 +1,237 @@
+
+	// emulationVersion can only be 1.{binaryMinor-1}...1.{binaryMinor}.
+	maxEmuVer := binaryVersion
+	minEmuVer := binaryVersion.SubtractMinor(1)
+	// TODO: remove in 1.32
+	// emulationVersion is introduced in 1.31, so it cannot be lower than that.
+	// binaryVersion could be lower than 1.31 in tests. So we are only checking 1.31.
+	if binaryVersion.EqualTo(version.MajorMinor(1, 31)) {
+		minEmuVer = version.MajorMinor(1, 31)
+	}
+	if emulationVersion.GreaterThan(maxEmuVer) || emulationVersion.LessThan(minEmuVer) {
+		errs = append(errs, fmt.Errorf("emulation version %s is not between [%s, %s]", emulationVersion.String(), minEmuVer.String(), maxEmuVer.String()))
+	}
+	// minCompatibilityVersion can only be 1.{binaryMinor-1} for alpha.
+	maxCompVer := binaryVersion.SubtractMinor(1)
+	minCompVer := binaryVersion.SubtractMinor(1)
+	if minCompatibilityVersion.GreaterThan(maxCompVer) || minCompatibilityVersion.LessThan(minCompVer) {
+		errs = append(errs, fmt.Errorf("minCompatibilityVersion version %s is not between [%s, %s]", minCompatibilityVersion.String(), minCompVer.String(), maxCompVer.String()))
+	}
+	return errs
+}
+
+// AddFlags adds the "{prefix}-emulated-version" to the flagset.
+func (m *effectiveVersion) AddFlags(fs *pflag.FlagSet, prefix string) {
+	if m == nil {
+		return
+	}
+	if len(prefix) > 0 && !strings.HasSuffix(prefix, "-") {
+		prefix += "-"
+	}
+	fs.Var(&m.emulationVersion, prefix+"emulated-version", ""+
+		"The version the K8s component emulates its capabilities (APIs, features, ...) of.\n"+
+		"If set, the component will emulate the behavior of this version instead of the underlying binary version.\n"+
+		"Any capabilities present in the binary version that were introduced after the emulated version will be unavailable and any capabilities removed after the emulated version will be available.\n"+
+		"This flag applies only to component capabilities, and does not disable bug fixes and performance improvements present in the binary version.\n"+
+		"Defaults to the binary version. The value should be between 1.{binaryMinorVersion-1} and 1.{binaryMinorVersion}.\n"+
+		"Format could only be major.minor")
+}
+
+func NewEffectiveVersion(binaryVer string) MutableEffectiveVersion {
+	effective := &effectiveVersion{}
+	binaryVersion := version.MustParse(binaryVer)
+	compatVersion := binaryVersion.SubtractMinor(1)
+	effective.Set(binaryVersion, binaryVersion, compatVersion)
+	return effective
+}
+
+// DefaultBuildEffectiveVersion returns the MutableEffectiveVersion based on the
+// current build information.
+func DefaultBuildEffectiveVersion() MutableEffectiveVersion {
+	verInfo := baseversion.Get()
+	ver := NewEffectiveVersion(verInfo.String())
+	if ver.BinaryVersion().Major() == 0 && ver.BinaryVersion().Minor() == 0 {
+		ver = DefaultKubeEffectiveVersion()
+	}
+	return ver
+}
+
+// DefaultKubeEffectiveVersion returns the MutableEffectiveVersion based on the
+// latest K8s release.
+// Should update for each minor release!
```

Comment:
how will we ensure this? perhaps an e2e tests that starts failing when the first beta is declared, but tolerates either N or N-1 with alpha? That's how we handled the "remove beta APIs after N releases"

Reply:
We will probably add a verify script to do that, tracked in https://github.com/kubernetes/kubernetes/issues/125032.